### PR TITLE
Set core version in one place

### PR DESCRIPTION
**Pull request by micbou**
Improve #171 by using cmake to build `ycm_core` with the version in the `CORE_VERSION` file.

Even if the preprocessor variable is defined by cmake, I am still defining it in `versioning.h` (if not already defined) with a bogus value to avoid a diagnostic error when editing the `versioning.cpp` file.

I could not find a good place to put the test so I created a new file for it.

### CODE REVIEW

File: cpp/ycm/versioning.h [hunk 1]
```diff
@@ -15,8 +15,17 @@
 // You should have received a copy of the GNU General Public License
 // along with YouCompleteMe.  If not, see <http://www.gnu.org/licenses/>.
 
+#ifndef VERSIONING_H_EEJUU0AH
+#define VERSIONING_H_EEJUU0AH
+
+#if !defined( YCMD_CORE_VERSION )
+  #define YCMD_CORE_VERSION 0
```
Comment (Valloric) [inline cpp/ycm/versioning.h]: Add a comment above this stating that the true value is set in a compiler command-line flag, and that we do so in the CMakeLists.txt file.

**Comment by Valloric** [comment]
Thanks for the PR! 

This is pretty nice. I added a minor inline comment; when that's resolved, this will be ready to merge.

**Comment by micbou** [comment]
I added the comment.

**Comment by Valloric** [comment]
Thanks!

@homu r+

@r4nt This will need special care when you pull it in.

**Pull request merged**